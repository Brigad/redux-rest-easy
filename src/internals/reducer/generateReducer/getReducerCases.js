import getInfosFromActionType from '../../utils/getInfosFromActionType';
import getResourceIdsByResourceNameFromNormalizedPayload from '../../utils/getResourceIdsByResourceNameFromNormalizedPayload';
import {
  computeNewResolversHashes,
  resetResourceResolversHashes,
} from '../../utils/resolversHashes';
import shallowMergeResources from '../../utils/shallowMergeResources';

const REDUCER_CASES = {
  REQUEST: (state, { type, url: normalizedURL, resourceId }) => ({
    ...state,
    requests: {
      ...(state.requests || {}),
      [normalizedURL]: {
        ...(state.requests && state.requests[normalizedURL]
          ? state.requests[normalizedURL]
          : {}),
        resourceName: getInfosFromActionType(type).resourceName,
        resourceId,
        startedAt: new Date().toISOString(),
        endedAt: null,
        expireAt: null,
        hasSucceeded:
          state.requests && state.requests[normalizedURL]
            ? state.requests[normalizedURL].hasSucceeded
            : false,
        hasFailed:
          state.requests && state.requests[normalizedURL]
            ? state.requests[normalizedURL].hasFailed
            : false,
        didInvalidate:
          state.requests && state.requests[normalizedURL]
            ? state.requests[normalizedURL].didInvalidate
            : false,
        fromCache:
          state.requests && state.requests[normalizedURL]
            ? state.requests[normalizedURL].fromCache
            : false,
      },
    },
  }),
  RECEIVE: (
    state,
    {
      type,
      url: normalizedURL,
      payload: normalizedPayload,
      principalResourceIds,
      metadata,
      cacheLifetime = 0,
    },
  ) => {
    const { resourceName } = getInfosFromActionType(type);
    const newStateOldHashes = {
      ...state,
      requests: {
        ...(state.requests || {}),
        [normalizedURL]: {
          ...(state.requests && state.requests[normalizedURL]
            ? state.requests[normalizedURL]
            : {}),
          endedAt: new Date().toISOString(),
          expireAt:
            cacheLifetime !== Infinity
              ? new Date(
                  new Date().getTime() + cacheLifetime * 1000,
                ).toISOString()
              : 'never',
          hasSucceeded: true,
          hasFailed: false,
          didInvalidate: false,
          fromCache: false,
          payloadIds: getResourceIdsByResourceNameFromNormalizedPayload(
            resourceName,
            normalizedPayload,
            principalResourceIds,
          ),
          metadata,
        },
      },
      resources: shallowMergeResources(state, normalizedPayload),
    };

    const newState = {
      ...newStateOldHashes,
      resolversHashes: computeNewResolversHashes(
        newStateOldHashes,
        resourceName,
        normalizedURL,
        normalizedPayload,
        principalResourceIds,
      ),
    };

    return newState;
  },
  FAIL: (state, { url: normalizedURL }) => ({
    ...state,
    requests: {
      ...(state.requests || {}),
      [normalizedURL]: {
        ...(state.requests && state.requests[normalizedURL]
          ? state.requests[normalizedURL]
          : {}),
        endedAt: new Date().toISOString(),
        hasSucceeded: false,
        hasFailed: true,
      },
    },
  }),
  RECEIVE_FROM_CACHE: (
    state,
    {
      type,
      url: normalizedURL,
      resourceId,
      payload: normalizedPayload,
      principalResourceIds,
      metadata,
      cacheLifetime = 0,
    },
  ) => {
    const { resourceName } = getInfosFromActionType(type);
    const newStateOldHashes = {
      ...state,
      requests: {
        ...(state.requests || {}),
        [normalizedURL]: {
          ...(state.requests && state.requests[normalizedURL]
            ? state.requests[normalizedURL]
            : {}),
          resourceName: getInfosFromActionType(type).resourceName,
          resourceId,
          startedAt: new Date().toISOString(),
          endedAt: new Date().toISOString(),
          expireAt:
            cacheLifetime !== Infinity
              ? new Date(
                  new Date().getTime() + cacheLifetime * 1000,
                ).toISOString()
              : 'never',
          hasSucceeded: true,
          hasFailed: false,
          didInvalidate: false,
          fromCache: true,
          payloadIds: getResourceIdsByResourceNameFromNormalizedPayload(
            resourceName,
            normalizedPayload,
            principalResourceIds,
          ),
          metadata,
        },
      },
      resources: shallowMergeResources(state, normalizedPayload),
    };

    const newState = {
      ...newStateOldHashes,
      resolversHashes: computeNewResolversHashes(
        newStateOldHashes,
        resourceName,
        normalizedURL,
        normalizedPayload,
        principalResourceIds,
      ),
    };

    return newState;
  },
  INVALIDATE_RESOURCE: (state, { resourceName }) => ({
    ...state,
    requests: {
      ...Object.entries(state.requests || {}).reduce(
        (prev, [key, request]) => ({
          ...prev,
          [key]:
            !request.didInvalidate
            && (request.resourceName === resourceName
              || (request.payloadIds && request.payloadIds[resourceName]))
              ? {
                  ...request,
                  didInvalidate: true,
                }
              : request,
        }),
        {},
      ),
    },
  }),
  INVALIDATE_ID: (state, { resourceName, resourceId }) => ({
    ...state,
    requests: {
      ...Object.entries(state.requests || {}).reduce(
        (prev, [key, request]) => ({
          ...prev,
          [key]:
            !request.didInvalidate
            && ((request.resourceName === resourceName
              && request.resourceId === resourceId)
              || (request.payloadIds
                && request.payloadIds[resourceName]
                && request.payloadIds[resourceName]
                  .map(item => item.toString())
                  .includes(resourceId.toString())))
              ? {
                  ...request,
                  didInvalidate: true,
                }
              : request,
        }),
        {},
      ),
    },
  }),
  INVALIDATE_REQUEST: (state, { url: normalizedURL }) => ({
    ...state,
    requests: {
      ...(state.requests || {}),
      [normalizedURL]: {
        ...(state.requests && state.requests[normalizedURL]
          ? state.requests[normalizedURL]
          : {}),
        didInvalidate: true,
      },
    },
  }),
  RESET_RESOURCE: (state, { resourceName }) => {
    const newStateOldHashes = {
      ...state,
      requests: {
        ...Object.entries(state.requests || {})
          .filter(
            ([, { resourceName: name, payloadIds }]) =>
              resourceName !== name
              && (!payloadIds || !Object.keys(payloadIds).includes(resourceName)),
          )
          .reduce(
            (prev, [key, request]) => ({
              ...prev,
              [key]: request,
            }),
            {},
          ),
      },
      resources: {
        ...(state.resources || {}),
        [resourceName]: undefined,
      },
    };

    const newState = {
      ...newStateOldHashes,
      resolversHashes: resetResourceResolversHashes(
        newStateOldHashes,
        resourceName,
      ),
    };

    return newState;
  },
  RESET_ALL: () => {
    const newState = {};

    return newState;
  },
};

const getReducerCases = () => REDUCER_CASES;

export default getReducerCases;
