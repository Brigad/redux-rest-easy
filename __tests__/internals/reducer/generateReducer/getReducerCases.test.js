import mockdate from 'mockdate';
import moment from 'moment';

import getReducerCases from '../../../../src/internals/reducer/generateReducer/getReducerCases';

const CACHE_LIFETIME = 0;
const RESOURCE_NAME = 'fruits';
const RESOURCE_ID = 2;
const ACTION_NAME = 'eat';
const NORMALIZED_URL = 'eat:https://api.co/fruits';
const MOCK_DATE = moment
  .utc()
  .year(2017)
  .month(0)
  .date(1)
  .startOf('day');

const INITIAL_STATE_EMPTY = {};
const INITIAL_STATE_FULL = {
  requests: {
    [NORMALIZED_URL]: {
      resourceName: RESOURCE_NAME,
      didInvalidate: true,
    },
    someURL: {
      resourceName: RESOURCE_NAME,
      resourceId: RESOURCE_ID,
      didInvalidate: false,
    },
    someOtherURL: {
      didInvalidate: false,
      payloadIds: {
        [RESOURCE_NAME]: [1],
      },
    },
    anotherURL: {
      didInvalidate: false,
      payloadIds: {
        [RESOURCE_NAME]: [1, RESOURCE_ID, 3],
        vegetables: [RESOURCE_ID],
      },
    },
    yetAnotherURL: {
      didInvalidate: false,
      payloadIds: {
        vegetables: [1],
        seeds: [RESOURCE_ID],
      },
    },
  },
  resources: {
    [RESOURCE_NAME]: {
      1: 'banana',
      [RESOURCE_ID]: 'cherry',
      3: 'apple',
    },
    vegetables: {
      1: 'potato',
      [RESOURCE_ID]: 'leek',
    },
    seeds: {
      [RESOURCE_ID]: 'pumpkin',
    },
  },
};

const REDUCER_CASES = getReducerCases();

describe('getReducerCases', () => {
  afterEach(() => {
    mockdate.reset();
  });

  test('reducer cases keys', () => {
    expect(REDUCER_CASES).toMatchSnapshot();
  });

  test('REQUEST', () => {
    mockdate.set(MOCK_DATE);

    const action = {
      type: `@@rest-easy/${RESOURCE_NAME}/${ACTION_NAME}/REQUEST`,
      url: NORMALIZED_URL,
      resourceId: 2,
    };

    expect(
      REDUCER_CASES.REQUEST(INITIAL_STATE_EMPTY, action),
    ).toMatchSnapshot();
  });

  test('RECEIVE', () => {
    mockdate.set(MOCK_DATE);

    const action = {
      type: `@@rest-easy/${RESOURCE_NAME}/${ACTION_NAME}/RECEIVE`,
      url: NORMALIZED_URL,
      payload: {
        fruits: {
          1: 'banana',
          2: 'cherry',
          3: 'apple',
        },
        animals: {
          1: 'worm',
        },
      },
      principalResourceIds: ['2', '1', '3'],
      cacheLifetime: CACHE_LIFETIME,
    };

    expect(
      REDUCER_CASES.RECEIVE(INITIAL_STATE_EMPTY, action),
    ).toMatchSnapshot();
  });

  test('FAIL', () => {
    mockdate.set(MOCK_DATE);

    const action = {
      url: NORMALIZED_URL,
    };

    expect(REDUCER_CASES.FAIL(INITIAL_STATE_EMPTY, action)).toMatchSnapshot();
  });

  test('RECEIVE_FROM_CACHE', () => {
    mockdate.set(MOCK_DATE);

    const action = {
      type: `@@rest-easy/${RESOURCE_NAME}/${ACTION_NAME}/RECEIVE_FROM_CACHE`,
      url: NORMALIZED_URL,
      resourceId: 2,
      payload: {
        fruits: {
          1: null,
        },
      },
      principalResourceIds: ['1'],
      cacheLifetime: CACHE_LIFETIME,
    };

    expect(
      REDUCER_CASES.RECEIVE_FROM_CACHE(INITIAL_STATE_EMPTY, action),
    ).toMatchSnapshot();
  });

  test('INVALIDATE_RESOURCE', () => {
    const action = {
      resourceName: RESOURCE_NAME,
    };

    const result = REDUCER_CASES.INVALIDATE_RESOURCE(
      INITIAL_STATE_FULL,
      action,
    );

    expect(result).toMatchSnapshot();
    expect(result.requests.anotherURL).not.toBe(
      INITIAL_STATE_FULL.requests.anotherURL,
    );
    expect(result.requests.yetAnotherURL).toBe(
      INITIAL_STATE_FULL.requests.yetAnotherURL,
    );
  });

  test('INVALIDATE_ID', () => {
    const action = {
      resourceName: 'vegetables',
      resourceId: RESOURCE_ID,
    };

    const result = REDUCER_CASES.INVALIDATE_ID(INITIAL_STATE_FULL, action);

    expect(result).toMatchSnapshot();
    expect(result.requests.anotherURL).not.toBe(
      INITIAL_STATE_FULL.requests.anotherURL,
    );
    expect(result.requests.yetAnotherURL).toBe(
      INITIAL_STATE_FULL.requests.yetAnotherURL,
    );
  });

  test('INVALIDATE_REQUEST', () => {
    const action = {
      url: NORMALIZED_URL,
    };

    const result = REDUCER_CASES.INVALIDATE_REQUEST(INITIAL_STATE_FULL, action);

    expect(result).toMatchSnapshot();
    expect(result.requests[NORMALIZED_URL]).not.toBe(
      INITIAL_STATE_FULL.requests[NORMALIZED_URL],
    );
    expect(result.requests.anotherURL).toBe(
      INITIAL_STATE_FULL.requests.anotherURL,
    );
  });

  test('RESET_RESOURCE', () => {
    const action = {
      resourceName: RESOURCE_NAME,
    };

    expect(
      REDUCER_CASES.RESET_RESOURCE(INITIAL_STATE_FULL, action),
    ).toMatchSnapshot();
  });

  test('RESET_ALL', () => {
    const action = {};

    expect(
      REDUCER_CASES.RESET_ALL(INITIAL_STATE_FULL, action),
    ).toMatchSnapshot();
  });
});
