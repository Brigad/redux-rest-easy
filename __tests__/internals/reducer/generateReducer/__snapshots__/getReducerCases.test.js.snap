// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getReducerCases FAIL 1`] = `
Object {
  "requests": Object {
    "eat:https://api.co/fruits": Object {
      "endedAt": "2017-01-01T00:00:00.000Z",
      "hasFailed": true,
      "hasSucceeded": false,
    },
  },
}
`;

exports[`getReducerCases INVALIDATE_ID 1`] = `
Object {
  "requests": Object {
    "anotherURL": Object {
      "didInvalidate": true,
      "payloadIds": Object {
        "fruits": Array [
          1,
          2,
          3,
        ],
        "vegetables": Array [
          2,
        ],
      },
    },
    "eat:https://api.co/fruits": Object {
      "didInvalidate": true,
      "resourceName": "fruits",
    },
    "someOtherURL": Object {
      "didInvalidate": false,
      "payloadIds": Object {
        "fruits": Array [
          1,
        ],
      },
    },
    "someURL": Object {
      "didInvalidate": false,
      "resourceId": 2,
      "resourceName": "fruits",
    },
    "yetAnotherURL": Object {
      "didInvalidate": false,
      "payloadIds": Object {
        "seeds": Array [
          2,
        ],
        "vegetables": Array [
          1,
        ],
      },
    },
  },
  "resources": Object {
    "fruits": Object {
      "1": "banana",
      "2": "cherry",
      "3": "apple",
    },
    "seeds": Object {
      "2": "pumpkin",
    },
    "vegetables": Object {
      "1": "potato",
      "2": "leek",
    },
  },
}
`;

exports[`getReducerCases INVALIDATE_REQUEST 1`] = `
Object {
  "requests": Object {
    "anotherURL": Object {
      "didInvalidate": false,
      "payloadIds": Object {
        "fruits": Array [
          1,
          2,
          3,
        ],
        "vegetables": Array [
          2,
        ],
      },
    },
    "eat:https://api.co/fruits": Object {
      "didInvalidate": true,
      "resourceName": "fruits",
    },
    "someOtherURL": Object {
      "didInvalidate": false,
      "payloadIds": Object {
        "fruits": Array [
          1,
        ],
      },
    },
    "someURL": Object {
      "didInvalidate": false,
      "resourceId": 2,
      "resourceName": "fruits",
    },
    "yetAnotherURL": Object {
      "didInvalidate": false,
      "payloadIds": Object {
        "seeds": Array [
          2,
        ],
        "vegetables": Array [
          1,
        ],
      },
    },
  },
  "resources": Object {
    "fruits": Object {
      "1": "banana",
      "2": "cherry",
      "3": "apple",
    },
    "seeds": Object {
      "2": "pumpkin",
    },
    "vegetables": Object {
      "1": "potato",
      "2": "leek",
    },
  },
}
`;

exports[`getReducerCases INVALIDATE_RESOURCE 1`] = `
Object {
  "requests": Object {
    "anotherURL": Object {
      "didInvalidate": true,
      "payloadIds": Object {
        "fruits": Array [
          1,
          2,
          3,
        ],
        "vegetables": Array [
          2,
        ],
      },
    },
    "eat:https://api.co/fruits": Object {
      "didInvalidate": true,
      "resourceName": "fruits",
    },
    "someOtherURL": Object {
      "didInvalidate": true,
      "payloadIds": Object {
        "fruits": Array [
          1,
        ],
      },
    },
    "someURL": Object {
      "didInvalidate": true,
      "resourceId": 2,
      "resourceName": "fruits",
    },
    "yetAnotherURL": Object {
      "didInvalidate": false,
      "payloadIds": Object {
        "seeds": Array [
          2,
        ],
        "vegetables": Array [
          1,
        ],
      },
    },
  },
  "resources": Object {
    "fruits": Object {
      "1": "banana",
      "2": "cherry",
      "3": "apple",
    },
    "seeds": Object {
      "2": "pumpkin",
    },
    "vegetables": Object {
      "1": "potato",
      "2": "leek",
    },
  },
}
`;

exports[`getReducerCases RECEIVE 1`] = `
Object {
  "requests": Object {
    "eat:https://api.co/fruits": Object {
      "didInvalidate": false,
      "endedAt": "2017-01-01T00:00:00.000Z",
      "fromCache": false,
      "hasFailed": false,
      "hasSucceeded": true,
      "payloadIds": Object {
        "animals": Array [
          "1",
        ],
        "fruits": Array [
          "2",
          "1",
          "3",
        ],
      },
    },
  },
  "resolversHashes": Object {
    "requests": Object {
      "eat:https://api.co/fruits": Object {
        "animals": "7054a0150530f9b3dd67e1123b18684e",
        "fruits": "7c9540c48d3b4c3b7c9867b5e00315ef",
      },
    },
    "resources": Object {
      "_getResourcesHash": [Function],
      "fruits": "29978190b54cefaef1578d2192d02a52",
    },
  },
  "resources": Object {
    "animals": Object {
      "1": "worm",
    },
    "fruits": Object {
      "1": "banana",
      "2": "cherry",
      "3": "apple",
    },
  },
}
`;

exports[`getReducerCases RECEIVE_FROM_CACHE 1`] = `
Object {
  "requests": Object {
    "eat:https://api.co/fruits": Object {
      "didInvalidate": false,
      "endedAt": "2017-01-01T00:00:00.000Z",
      "fromCache": true,
      "hasFailed": false,
      "hasSucceeded": true,
      "payloadIds": Object {
        "fruits": Array [
          "1",
        ],
      },
      "resourceId": 2,
      "resourceName": "fruits",
      "startedAt": "2017-01-01T00:00:00.000Z",
    },
  },
  "resolversHashes": Object {
    "requests": Object {
      "eat:https://api.co/fruits": Object {
        "fruits": "7054a0150530f9b3dd67e1123b18684e",
      },
    },
    "resources": Object {
      "_getResourcesHash": [Function],
      "fruits": "761b5b7bf83a5eba0ae3f64c9ec28651",
    },
  },
  "resources": Object {
    "fruits": Object {
      "1": null,
    },
  },
}
`;

exports[`getReducerCases REQUEST 1`] = `
Object {
  "requests": Object {
    "eat:https://api.co/fruits": Object {
      "didInvalidate": false,
      "endedAt": null,
      "fromCache": false,
      "hasFailed": false,
      "hasSucceeded": false,
      "resourceId": 2,
      "resourceName": "fruits",
      "startedAt": "2017-01-01T00:00:00.000Z",
    },
  },
}
`;

exports[`getReducerCases RESET_ALL 1`] = `Object {}`;

exports[`getReducerCases RESET_RESOURCE 1`] = `
Object {
  "requests": Object {
    "yetAnotherURL": Object {
      "didInvalidate": false,
      "payloadIds": Object {
        "seeds": Array [
          2,
        ],
        "vegetables": Array [
          1,
        ],
      },
    },
  },
  "resolversHashes": Object {
    "requests": Object {},
    "resources": Object {
      "fruits": "cf252fdcd0c577918a558e3d5869780d",
    },
  },
  "resources": Object {
    "fruits": undefined,
    "seeds": Object {
      "2": "pumpkin",
    },
    "vegetables": Object {
      "1": "potato",
      "2": "leek",
    },
  },
}
`;

exports[`getReducerCases reducer cases keys 1`] = `
Object {
  "FAIL": [Function],
  "INVALIDATE_ID": [Function],
  "INVALIDATE_REQUEST": [Function],
  "INVALIDATE_RESOURCE": [Function],
  "RECEIVE": [Function],
  "RECEIVE_FROM_CACHE": [Function],
  "REQUEST": [Function],
  "RESET_ALL": [Function],
  "RESET_RESOURCE": [Function],
}
`;
